src/Hopper/Internal/LoweredCore/ANF.hs

```diff
@@ -0,0 +1,477 @@
+               -> BindingStack
+trackVariables (AnfBinding refs) vars stack =
+  stack & _head.levelRefs %~ addRefs
+  where
+    addRefs m = foldl' (flip $ uncurry Map.insert) m $ zip refs vars
+trackVariables TermBinding vars stack =
+  emptyIndirectionLevel vars : stack
+
+-- | Updates the top of the 'BindingStack' to open a new 'AnfRef' for an
+-- introduced 'AnfLet', or adds a new 'BindingLevel' for an existing 'Term'
+-- 'Let'.
+trackBinding :: Binding -> BindingStack -> BindingStack
+trackBinding (AnfBinding refs) stack = stack & _head %~ updateLevel
+  where
+    updateLevel = (levelRefs                              %~ addRefs)
+                . (levelRefs.mapped.localNameless.lnDepth +~ 1)
+                . (levelIntros                            +~ 1)
+    addRefs m = foldl' (\m' ref -> Map.insert ref v0 m') m refs
+trackBinding TermBinding stack = emptyLevel:stack
+
+-- | Stops tracking the provided 'AnfRef's in the top 'BindingLevel' of
+-- 'BindingStack'.
+dropRefs :: Foldable t => t AnfRef -> BindingStack -> BindingStack
+dropRefs refs stack = stack & _head.levelRefs %~ deleteRefs
+  where
+    deleteRefs m = foldl' (flip Map.delete) m refs
+
+-- | Resolves 'Variables' in 'BindingLevel' for the provided 'AnfRef's. Assumes
+-- the refs are all present in the 'Map' of the 'BindingLevel'.
+resolveRefs :: (Foldable t, Functor t)
+            => t AnfRef
+            -> BindingStack
+            -> t Variable
+resolveRefs refs stack = (varMap Map.!) <$> refs
+  where
+    varMap = fromMaybe (error "unexpected binding stack underrun") $
+      firstOf (_head.levelRefs) stack
+
+-- | This is what we currently use to represent a deferred transformation to
+-- 'BindingStack' (from the use of 'trackVariables' or 'trackBinding') for the
+-- 'convertWithCont' callee to invoke. We typically call it immediately in the
+-- continuation, except in the 'Let' case, where we need to roll back the stack
+-- before applying the transform.
+type StackTransform = BindingStack -> BindingStack
+
+-- | Lowering a sequence of "sibling" 'Term's, and providing 'Variable's for the
+-- lowering of the rest of the program.
+convertToVars :: [Term]
+              -- ^ A sequence of 'Terms' to lower in order, e.g. for prim or
+              -- function or thunk application, or returning multiple values.
+              -> ([Variable] -> StackTransform -> LoweringM Anf)
+              -- ^ Continuation for synthesizing the lowering for the rest of
+              -- the program from 'Variable's for each of the terms and a
+              -- 'StackTransform' for cleaning up the 'AnfRef's that tracked the
+              -- 'Variable's.
+              -> LoweringM Anf
+convertToVars terms synthesize = do
+  refs <- replicateM (length terms) allocRef
+
+  -- NOTE: To see how we could simplify this function, see notes on the 'Let'
+  -- case of 'convertWithCont' regarding the possibility of using @ContT@. The
+  -- fold in this function would become first-order.
+  id &
```

Comment:
I'd probably call the second arg of `foldr` something like `innerContinuation`. It's probably a good idea to name this.

I decided against `\(t, ref) nextK track ->` to make it clear that we are returning/building up a continuation -- something that takes `track` and returns a lowering action for the rest of the program. Also so people don't get confused when they see `foldr`'s function (`:: a -> b -> b`) taking more than two args.